### PR TITLE
[IMP] Chart: use first row as headers

### PR DESCRIPTION
**Pull request by yunchen4**
## Description:

This PR changes the current option "Datasets include title" to "Use row/col x as headers", which is more similar to the functionality in GSheets. It will exclude the first row/col in the datasets/labels when creating the chart.

Odoo task ID : [3099995](https://www.odoo.com/web#id=3099995&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts
Comment (pro-odoo) [inline src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:23]: We already have a type for that: https://github.com/odoo/o-spreadsheet/blob/master/src/types/misc.ts#L250

File: src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts
Comment (pro-odoo) [inline src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:108]: why `typeof` ?

File: src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts
Comment (pro-odoo) [inline src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:112]: `as const` should not be necessary

File: src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts
Comment (pro-odoo) [inline src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:116]:
```
const getters = this.env.model.getters;
const sheetId = getters.getActiveSheetId();
```

File: tests/__snapshots__/xlsx_export.test.ts.snap
Comment (pro-odoo) [inline tests/__snapshots__/xlsx_export.test.ts.snap:90]: Is it really correct ? I'm not sure...

File: tests/collaborative/collaborative_sheet_manipulations.test.ts
Comment (pro-odoo) [inline tests/collaborative/collaborative_sheet_manipulations.test.ts:582]: Seems useless diff (actually the whole file)

File: tests/plugins/chart/basic_chart.test.ts
Comment (pro-odoo) [inline tests/plugins/chart/basic_chart.test.ts:79]:
```suggestion
        labelRange: "Sheet1!A2:A4",
        dataSetsHaveTitle: true,
```
Considering the test data, it should be it, right ? Or I missed something ?

**Review by pro-odoo — CHANGES_REQUESTED** [review]
Little issue with excel import/export:
* Download demo sheet as xlsx
* Import it
 => The line graph is not correct, the range has change and the row header is not ticked anymore.

File: tests/plugins/chart/basic_chart.test.ts
Comment (yunchen4) [reply tests/plugins/chart/basic_chart.test.ts:79]: > Considering the test data, it should be it, right ? Or I missed something ?

@pro-odoo Previously when the datasets have titles, the category XC will be automatically decreased by one, so if the dataset range is `B1:B4` including title, the label range will be changed into `A2:B4`. 
Now I cancel this automatic behavior and make `use row x as headers` to exclude the label header.

File: src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts
Comment (LucasLefevre) [inline src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:42]: I don't think we need to support "column".
It's enough for most cases.
Keep it simple :)

File: src/xlsx/conversion/figure_conversion.ts
Comment (yunchen4) [reply src/xlsx/conversion/figure_conversion.ts:55]: @LucasLefevre Because now we exclude the whole row rather than a cell when we ticked `dataset has title`, it will also influence the exporting .xlsx file. And also when we import .xlsx file we also want to automatically tick "use row x as headers" if applicable, so part of code are also about it.

File: src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts
Comment (yunchen4) [inline src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:43]: I still keep the state here because two `dispatchResult` are also states. If I don't use state for header index, when the dataset/label is actually invalid, the states of dispatch results are not updated immediately (which leads to valid judgement), so `calculateHeadPosition()` will throw error and the whole app crashes. I don't know whether there is any way better than states 🤔

**Review by LucasLefevre — COMMENTED** [review]
partial review from last week :)

File: src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts
Comment (yunchen4) [reply src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:40]: @LucasLefevre I remembered I tried to calculate it dynamically, but the problem is the label range and the data set need to be valid first, then we can do the calculation. However, the dispatch results which determine the validness are stored in the state which is not updated immediately (if it works like React in which states are updated in batch, or at least when calling `calculateHeaderPosition()` they are not updated, even when keeping the first lines of code in `calculateHeaderPosition()` which are used to return early if invalid), and it will throw errors to cause the whole app crash. Using `t-if` cannot avoid it either. 
Hence I store `headerPosition` and only update it in setup and after the dataset and label range are confirmed, to be sure that it will not cause error and I think it is enough to update the header position only after confirmation. 
Do you have any idea to make it calculate dynamically without throwing errors related to invalid dataset/label range?

File: src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts
Comment (rrahir) [reply src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:109]: You could return `undefined` when the dataset/label is invalid, and use that specific value to decide wheter or not  to display the checkbox ?

Comment (yunchen4) [reply src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:109]: @rrahir I can directly use `isDatasetInvalid` and `isLabelInvalid` to do the conditional rendering, but the problem is when calling `calculateHeaderPosition()` these two invalid indicators are not updated on time (see my comment in the last conversation).

**Review by rrahir — CHANGES_REQUESTED** [review]

File: src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts
Comment (yunchen4) [reply src/components/side_panel/chart/line_bar_pie_panel/config_panel.ts:109]: @rrahir Maybe I can push the commit where it is (I think) dynamically compute `headerPosition` but it crashes when the dataset input is invalid? I understand the flow but there might be something I put wrongly : ( But I now know that I need to return something like `undefined` to differentiate it. 
> here is the gist of the diff that I applied it seems to work like a charm.
What gist?

File: src/helpers/figures/charts/chart_common.ts
Comment (yunchen4) [reply src/helpers/figures/charts/chart_common.ts:282]: @rrahir I wait until #1990 is merged?

File: src/helpers/figures/charts/chart_common.ts
Comment (yunchen4) [reply src/helpers/figures/charts/chart_common.ts:284]: I saw in other similar functions they have this step so I did it

File: src/helpers/figures/charts/chart_common.ts
Comment (rrahir) [reply src/helpers/figures/charts/chart_common.ts:282]: it is, so you can rebase and use `splitReference`  ;-)

File: src/helpers/figures/charts/chart_common.ts
Comment (rrahir) [inline src/helpers/figures/charts/chart_common.ts:285]: you can feed the `xc` from plitReference` here :)

File: src/helpers/figures/charts/chart_common.ts
Comment (rrahir) [inline src/helpers/figures/charts/chart_common.ts:308]: same comment as above

File: src/helpers/figures/charts/chart_common.ts
Comment (yunchen4) [reply src/helpers/figures/charts/chart_common.ts:284]: @rrahir I checked and you're right, I'll delete the replacement of dollar sign. But what do you mean by fixed rows/cols? What's the use case? 🤔

File: src/helpers/figures/charts/chart_common.ts
Comment (rrahir) [inline src/helpers/figures/charts/chart_common.ts:282]: this causes an issue, we have to discuss this tomorrow. THe sheetName yields does not have quotes its the pure sheet name. So when you combine your sheetName and dataXC later on, you should check taht hte quotes are present if  necessary. a part of this logic is present and is inadequately called `getComposerSheetName` so you can use  this ATM but it highlights the need to set some kind of guidelines on how to process sheetNames

File: src/xlsx/conversion/figure_conversion.ts
Comment (rrahir) [reply src/xlsx/conversion/figure_conversion.ts:55]: for the record, we are already converting excel stuff in the function (if you look at the function arguments)

we could argue that those helpers `convertExcelDatasetToSheetXC` and `convertExcelLabelToSheetXC` don't belong in chart common - they should be defined in the place where they are used -`figure_conversion.ts`. 

FUrthermore, both helpers seem to do the exact same thing (we only process the range of the ExcelChartdataset, and in the same way we do with the label) So it'd be nice to have a single function `convertExcelRangeToSHeetXC` or have both functions call a common helper.

File: demo/data.js
Comment (rrahir) [inline demo/data.js:189]: useless diff ?

**Review by rrahir — CHANGES_REQUESTED** [review]
Lots of changes done in the label ranges which are not necessary - those should be rolled back.

I think what we are missing (unless I didn't see the test, if so I apologize) a test that checks that exporting a chart with the label range containing the title row or not leads to the same exported xlsx.
e.g.
``` javascript
{
    labelRange: A1:A3,
    dataSets: [B1:B3],
    dataSetsHaveTitle: true
}
```
genetares the same xlsx as

``` javascript
{
    labelRange: A2:A3,
    dataSets: [B1:B3],
    dataSetsHaveTitle: true
}
```

File: tests/xlsx_export.test.ts
Comment (yunchen4) [inline tests/xlsx_export.test.ts:888]: I doubted those `C12` were typos.

**Pull request closed without merging**
